### ISSUE TITLE
tremendous dependency size

### ISSUE THREAD

**Issue by mechanoid** [tapjs/tapjs#333]
After recommended tap over tape to a colleague, he pointed out the big difference in file size due to dependencies.

As a comparison `node-tape` with dependencies installed weighs in at about 2.4MB disk usage (which i consider too much for a minimal test framework), but looking at `node-tap` we end in 54MB disk usage. 

After digging deeper it looks like the main disk space is used for 2 things

1.) polyfilling (bluebird) with about 700kb
2.) code-coverage
** coveralls with about 6MB
** and nyc around the 45MB i guess

For 1.) i consider it a bad practice to include polyfils by default so in my opinion it should be a peerDependency, which people may drag in if it is necessary.

For 2.) i personally think that the mix of a testing utility and line-coverage is merely useless, especially when it is just a nice wrapper of starting the test util with another Cli-Interface.
So in my opinion it would be nice to make those dependencies also linkable via peerDependencies, 
so that they can be implemented if given and still remove their mandatory nature.

Overall i think tap is the much better test framework, but i can understand people moving back to tape (or other solutions) because of that. 

Thanks in advance, Falk

**Comment by Tokimon** [comment]
yes totally agree.
I love all the great features node-tap provides, but personally I hold back because of the different dependencies that could be avoided.

**Comment by ORESoftware** [comment]
Yeah I agree, I really don't know why it's not more common to store certain big dependencies (like babel, coveralls, istanbul/nyc, webpack) in a single global location. You'd have to take into account the version of the package and point to that version somehow, but IMO someone needs to work on this problem. Perhaps Yarn, the new package manager will think about doing something like this.

**Comment by mechanoid** [comment]
Hmm i don't think that this gets the point here. I like the isolation between the application folders, i just don't know, why a library like istanbul with an independent command line interface (nyc) has to be included in a "minimal" testing library.

**Comment by ORESoftware** [comment]
@mechanoid agree and disagree. a minimal testing library like this one probably has less business integrating with nyc/istanbul. But more powerful testing libraries like AVA, Mocha and my own (Suman) provide users with some real benefit by integrating with NYC/Istanbul/Babel etc. However, including NYC/Babel/Istanbul as dependencies makes the install HUGE. So we are stuck between a rock and hard place. You did not address my proposal - why doesn't NPM allow us to store all packages in one central location on any machine? Like Maven. I have never heard a good reason. All packages and all versions of packages can just go in one place. That will save disk space. I don't get it!

**Comment by Tokimon** [comment]
@mechanoid I see your point and there is definitely something to be done in that matter, but I guess it is put of scope for this testing tool ;)
For the sake of this matter I also believe that the coverage part should be optional to include and not forced upon you, even if I always (and encourage other to as well) include coverage reporting with my testing. I my case I am not necesarily using the CLI interface of TAP to load the files and then the automated coverage report is not working, thus making it a redundant inclusion and I have to do it manually anyway.

**Comment by mechanoid** [comment]
@ORESoftware at first i see a difference in libraries like tap/tape and ava, mocha, etc. in that the are absolutely not minimal. And a reason why I prefer libraries like tape is that they follow the unix philosophy of doing one thing well rather than doing many not that good. 
To the point on different dependency stores I think that it has for sure it good parts (saves lot of disk space for sure), but also its downsides. A thing that happened to me very often in e.g. ruby is, that you have to dig into your lockfile (or whatever your language hands to you) to have a clear state of what is installed. After found out, which version your app references you have to turn over to your central store and lookup that code when you want to check what a certain version does to hunt bugs down etc. An approach of storing app dependencies in the app folder means, that i have all dependant code in one place and nothing else, so if i want to check what a certain library does, i can head to it and start to manipulate it etc. without effecting other projects.

**Comment by mechanoid** [comment]
@tokimon i personally do not really believe in loc coverage, but yes i think if it would be optional it would make much more sense, but digging deeper to the tap code base I don't believe that it is that easy to remove that necessity.

**Comment by Tokimon** [comment]
@mechanoid I haven't looked deep into the code, but it seems strange that it is closely tied to the coverage report :-/

**Comment by isaacs** [comment]
This is an interesting question, and hits at the fundamental vision and goals of this tool.

Tap is **not** a "minimal test framework".  I don't know where you would have gotten that idea, frankly!  There's is literally nothing "minimal" about node-tap.  It has pretty much every feature one might want.  It's absolutely a "batteries-included" kind of thing, by explicit design.

The goal of this library is that you do `tap test/*.js` in your package.json file, and everything Just Works.  If you want coverage, you can throw a `--cov` on there, and get reasonable output.  If it's running in an environment with a coveralls repo token, it'll automatically ship the coverage data over there.  It detects if it's running in a child process, or a tty, and presents output that's optimized for each situation.  Like promises more than callbacks?  Tap both consumes promises, and test functions also return a promise.  It has aliases for all the assertion functions you could possibly imagine (are you imagining a new one? let's talk about it, I'll probably accept a PR!)

The goal is that it works in every version of node from 0.10 on up (because even though it's not "supported", many people use it in production, including friends and employees of mine).  When it's safe to drop 0.10 support (which will be a breaking change), it'll probably be worth dropping Bluebird.  But  since Promise support isn't native in 0.10, it uses a polyfill.  (And, at least in 0.12 and 4.x where I tested it, Bluebird is also faster than native promises, though that's a small factor since tap is typically I/O bound, and the speed of the Promise implementation is almost never relevant.)

Looking forward, though my time to spend on OSS has dwindled significantly in light of work and family concerns, the plan is to add support for buffered test output, and eventually use that to also enable asynchronous/parallel tests.  All the features!

As if to make the point even more obvious, try running `tap test/*.js -Rnyan`.  _There is a nyan cat reporter; your argument is invalid._ :D

So, since I do create many "minimal" libraries for various things, and have at many times advocated very publicly for building small modules and small things, why have I not designed tap to be minimal?  Why is it, instead, more accurately described as a _maximal_ test framework?

This comes down to trade-offs, and the role that developer tools like tap play in the workflow experience.  (Consider also, one would be hard pressed to call npm a "minimal" package manager!)  Dev tools should make the development experience easy and enjoyable.

For things that are shipped with an application into production, like the actual web framework that powers your application, or a reusable library, there's a big benefit to be had for being very minimal.  It makes debugging easier (and debugging production is fundamentally much harder than anything running locally!)  But for development time, I like to fly first class.  I want the bells and whistles, I want it to Just Work.  I code on a Macbook Pro for pretty much that exact reason.  (Though, that's a whole other conversation, as it's yet to be seen if Apple will keep catering to my use case in the future.)

Modularity is much more subtle than merely "smallness".  I recommend reading [On the Criteria To Be Used in Decomposing Systems into Modules](https://www.cs.umd.edu/class/spring2003/cmsc838p/Design/criteria.pdf).  Each "module" should encapsulate one decision, such that it's straightforward to replace.  Some decisions are indeed "minimal", and some are not.  Module implementations should be the minimum size required to take a specific stance on a given decision, _but no smaller_.

In tap's case, it encapsulates a bunch of opinions regarding the decision of "how to run software tests".  The resulting implementation also has, within itself, many other decisions that are themselves individually abstracted out into smaller (or, in nyc's case as mentioned here, not-so-smaller) modules of functionality.

The fact that it's so easy to switch from tap to tape (or to ava, or mocha, or whatever else), some of which _are_ certainly a lot more "minimal", indicates to me that this is an appropriate module boundary, and that minimal-ness is just one of several possible options for how one might decide to make this particular decision.  Tape is the easiest to switch to, and has the fewest features of any of them.  If that is your preference, use tape with my blessing!

In short, I'm definitely open to replacing Bluebird with native promises once 0.10 is no longer supported, and it's entirely possible that there are some ways to make nyc a bit leaner.  But I'm not likely to ever remove nyc as a dependency, or not have built-in coverage reporting in this library.

**Comment by ORESoftware** [comment]
@isaacs right, but why doesn't NPM just use the NPM cache to source modules instead of re-writing modules locally in every single project directory :) I 100% agree with you that it makes sense to include NYC with Node-Tap as a dependency, but the real concern is size on disk for many people. After 3 years working with Node.js, I still do not understand why NPM doesn't just put all versions of modules into one directory. I am missing something....maybe it has to do with locks on folders, or perhaps to avoid global file corruption (making problems only local to one project), etc. Is there anything you can point me to that serves as a definitive answer to why NPM uses local installation and doesn't point to ~/.npm or the like? People surely have the same versions of Babel and NYC/Istanbul and Express installed on their filesystems in 10 different places, I just don't see the point, ATM. Willing to be enlightened.

**Comment by ORESoftware** [comment]
In other words this thread isn't really about NYC being bundled with Node-Tap, it's convenient for both the library authors and users, to do so. What this thread is about is having to write NYC to the filesystem in the local project. I ultimate do not understand the need to do that.

**Comment by isaacs** [comment]
@ORESoftware Check out ied and pnpm.  That's a fine discussion to have, but this is not the place for it :)

**Comment by ORESoftware** [comment]
sure thing, are ied and pnpm repos or discussion channels? never heard of either of 'em, will google

**Comment by ORESoftware** [comment]
Ok, I see that `ied` and `pnpm` are NPM modules...to be honest, I don't know why @yarnpkg doesn't do something similar, maybe someone knows

**Comment by mechanoid** [comment]
hey @isaacs, thanks for this *huge* answer :)

Honestly I don't know where the minimal thought came from, if it was made up by my own assumption or if I got it as a "minimal" recommendation. On first reading your answer i thought "hey, just a moment, wasn't that stated in their docs", but no, it seems not :)

But i guess I was willingly believing in that, because I didn't get the point on nyc. Or at least in my opinion tap isn't far from being minimal in terms of "doing one thing, but doing it extraordinarily well", besides of that second thing. But yes, the point of keeping the tool backwards compatible is true and i think getting the dependencies optional without pain, may be a little bit challenging i guess. 

But still in the long run (so in 2025 i guess) i would love to rethink if nyc support as a peerDependency would make tap even better.

**Comment by isaacs** [comment]
@mechanoid Why `peerDependency`?  Do you mean `optionalDependency`?  A peer dependency would just install it at the top level instead of installing it underneath `tap`, and opens the door for dependency version conflicts (aka "dependency hell").  But a peerDependency is still _required_ to be there, so there's no change in disk size.

>  Or at least in my opinion tap isn't far from being minimal in terms of "doing one thing, but doing it extraordinarily well"

Thank you!  But you should know that "do one thing well" does _not_ mean "minimal"!  If the "one thing" is large, then it will sometimes require a lot of features and code to do well.  A way to think about "do one thing well" practically is: abstract away a single decision, be easily replaced, and effectively satisfy the most common use case by default.  The _size_ of the implementation is an orthogonal concern, sometimes relevant, sometimes not.  At best it's a heuristic, since programs that _don't_ limit the scope of their modularity tend to be larger, but the opposite is only true if you restrict your analysis to small problems.

**Comment by billiegoose** [comment]
> What this thread is about is having to write NYC to the filesystem in the local project. I ultimate do not understand the need to do that.
> I don't know why @yarnpkg doesn't do something similar, maybe someone knows

@oresoftware Having written a (only partially working) node package installer, I can think of one reason why. Intuitively, you'd assume you only need one copy of every module on a given computer. However, it is *unintuitively* hard to actually make a global, versioned, node package store work, due to `require`s... quirkiness.

But I and some much smarter people (@alexanderGugel and @zkochan) are indeed working on the problem! Join our discussion! It's scattered in issues in the ied repo mostly. It started here: https://github.com/alexanderGugel/ied/issues/182#issuecomment-263439223 I've been MIA for a week (swamped with work stuff) but we're currently concocting a module store specification that would allow multiple alternative installers (ied, pnpm, my half-baked attempt, hopefully future versions of yarn) to all use a single, system-wide installation store.

**Comment by ORESoftware** [comment]
@wmhilton yeah cool I am following the ied, pnpm, yarn stuff. So what I am hearing is that node core require function makes it hard to do package management without local modules, can you be more specific?

**Comment by mechanoid** [comment]
@isaacs oh darn, yes you are right, just mixed the terms up, yes for sure i meant optional dependencies. 

And i fully agree to the second part, thats one of the reasons i stick to the "extraordinarily well". Especially after digging a bit deeper into `tap`. So testing (and reporting inclusively nyan-cat ^^) is the one thing in my opinion (and i totally accept that this is a complex field, which inherently has the necessity of writing a lot code)  and coverage is another.

**Comment by ORESoftware** [comment]
@wmhilton my guess by what you are saying is that by default node require() looks in the local node_modules dir, and in order for it to look elsewhere, you'd have to use NODE_PATH. Are local modules a thing only because the CommonJS require function looks locally only?

**Comment by zkochan** [comment]
@ORESoftware `require()` always looks into the local node_modules dir. However, what it finds there, are symlinks. So the dependencies are physically only in one place on the disk. However, symlinks are adding it to all the dependent package's node_modules

**Comment by ORESoftware** [comment]
: ) So what came first, the decision for node's require() function to look locally only or NPM's decision to install locally only for a project? Sounds like the former. Maybe this is all because of CommonJS, or some decision way back when creating the require function in node.

**Comment by isaacs** [comment]
@ORESoftware Ryan and I developed both at the same time, around node 0.3.  The chicken and the egg evolved together from earlier forms.

**Comment by ORESoftware** [comment]
@isaacs can't argue with that - although I was under the impression that CommonJS preceded Node.js and that CommonJS was part of some other serverside JS thing. I am guessing that you and Ryan chose CommonJS as the module format, but I am wondering who originally drew up CommonJS (if it wasn't y'all)? I have to say that CommonJS has been pretty pleasant to work with.

**Comment by isaacs** [comment]
@ORESoftware I was involved with CommonJS (even back when it was called ServerJS).  Kris Kowal, Kris Zyp, Tom Robinson, and Kevin Dangoor were also quite active, as I recall.  The "Securable Module Format" was Kris Kowal's brainchild.  SSJS was a smaller community back then, though there were more technical players.  The module format was implemented in several slight variations in Narwhal, Helma/Ringo, and even CouchDB.  We implemented it in Node.js, and made a few node-specific changes.  Then CommonJS kinda died out, and we made even more changes.  Now those node-specific changes are chalked up as part of the "CommonJS Module Format", even though they had nothing to do with CommonJS, per se.

Also, the module format was one of a few dozen different specifications that the CommonJS movement spit out.  The `package.json` format started there as well, and eventually mutated into what npm uses today.  The first few Promises specs happened there, and a bunch of system, networking, http, and file system stuff, most of which is just in the ash heap of history at this point.

It's still weird to me that "CommonJS" means "`require()` and `node_modules`" now.  It'd be like if you said "I really like viewing webpages on Apple" to indicate that you enjoy using the Chrome browser (since it was originally based on WebKit, though is no longer).  Eh.  Languages are weird, I guess :)

**Comment by ORESoftware** [comment]
@isaacs you need write The Node.js History Book, already :) It's all very curious to me and I doubt I am the only one that wants to know! I am sure Ryan will want to co-author it at some point :)

You can call it The Node.js History Book Or How We Created The Biggest Open Source Ecosystem The World Has Ever Seen (And then they changed the module format to ES6)

**Comment by billiegoose** [comment]
@ORESoftware I think others have pretty much covered it, but here is a specific example. A lot of the pain comes from `require`'s fallback strategy: if dependency `foo` isn't in `node_modules/foo`, try `../node_modules/foo`, `../../node_modules/foo` and so on.

This strategy makes sense / works well for many, many cases:

1. Packages are != modules in node. Modules are files / dirs, and packages are a collection of these things. Strictly speaking packages do not even have to have a single `exports` interface; classic example would be `lodash` where you can require individual functions such as `var flatten = require('lodash/flatten')`. By using the `require` fallback strategy, package managers are able to work under the simplified assumption that any and all package dependencies can be placed in `lodash/node_modules/`. Otherwise, every sub directory within a module would need its own `node_modules` folder. (E.g. `myModule/node_modules`, `myModule/lib/node_modules`, `myModule/cli/node_modules` and you would end up having to have a `package.json` file *in every directory* which is ridiculous. The fallback strategy is an elegant solution here.

2. It allows for a (relatively) simple de-duplication algorithm. I.e. you can install packages in a plain ol' hierarchy as npm v2 did, then recursively "lift" modules from child `node_modules` directories into parent module `node_modules` directories. If lifting a package would conflict with a package already there, don't lift it, skip it and move on. This can greatly reduce the number of copies of modules that are required on disk.

But there are situations where module's behavior *depends* on this fallback strategy, and it becomes part of the API of node rather than an implementation detail:

1. *Peer dependencies* (may they live in infamy). One pattern of module development that is logical and quite common, is to have one core module (say `babel-core`) and lots of plugin modules (say `babel-*`). Obviously, you don't want to have each plugin include `babel` as a dependency, because that would mean if you install 12 babel plugins you'd install babel 12 times! Plus, each plugin might install a slightly different version of babel. The obvious solution? Don't list babel as a dependency, but still call `require('babel')` in the plugin code, and assume `babel` is available somehow. Here, the problem is usually solved by installing `babel` in the `node_modules` of some common-ancestor of all the babel plugins. But by doing this, we are *implicitly* relying on `require`'s fallback behavior, which is tied to the directory structure. If you replace one of your babel plugins with a symlink (e.g. replace `node_modules/babel-foo` with a symlink to `~/projects/babel-foo-fork`) how will `babel-foo-fork` be able to `require('babel')`? In previous node versions, symlinks were always resolved to real paths, so `require` would look in `~/projects/babel-foo-fork/node_modules`, `~/projects/node_modules`, `~/node_modules`.
2. In real-world projects, there are (unfortunately) packages that rely on the implicit fallback of `require`. If you are writing a module as part of a big project that uses the `chalk` package for coloring text, and you extract some part of big-project into big-project-plugin, you might forget to include `chalk` as a dependency of big-project-plugin. That's OK, because the only times that big-project-plugin are installed is when big-project is also installed, so `require('chalk')` is going to work, because `chalk` will be there, somewhere higher in the directory structure. You have the same problem as above - as soon as you try to physically relocate big-project-plugin to another directory, that implicit dependency is broken.
3. OK, you say. So simple, stop resolving symlinks to their real path. Node added that as a feature recently (`--preserve-symlinks`). Now you have an even *stranger* issue. Say you have a module you are working on (`~/projects/sir-log-a-lot`) and you are using it in every module of `big-project`. Now you have dozens of symlinks to `~/projects/sir-log-a-lot` in `big-project`. Now the dependencies load fine, but you run afoul of Node's module caching system. Node only executes a module once, the first time it is required. After that, future `require` calls just return the module.exports already in memory. However, the modules are cached based on their directory path. So now each time you require `sir-log-a-lot` you get an uncached, fresh copy. Which is fine except a) it wastes memory and b) two copies of `sir-log-a-lot` *no longer have shared state* so if you changed `sir-lot-a-lot`'s settings in one file, it would only affect that file's local copy of `sir-log-a-lot` rather than globally as intended.

These are all minor inconveniences if you are using trying to use symlinks as a means for substituting a module with a shared development module. In practice, you don't do that and just use `npm link` instead. However, if you wanted to completely change the directory layout on disk to do say `~/.store/packageName/version` ... there's no way to do this with symlinks that I am aware of. There ARE (probably?) ways to do it, but it is not as trivial as it should be. There is currently a pull request for a semantics where symlinks are preserved for `require`ing but resolved to their real path for caching... which seems like it could work at first glance. But the fact that someone had to patch `node` itself is a testament to how difficult the problem really is.

There may be other ways to work around this limitation. One way I'm currently considering is overriding `require` at runtime. Another is doing some chicanery with NODE_PATH, which takes precedence over the directory walking algorithm. But yeah, don't knock it until you try writing your own package manager. I've gained a lot of respect for @isaacs by trying to write a better `npm`. And I would definitely buy that book if he wrote one.
